File: 백준/Silver/15650. N과 M （2）/N과 M （2）.js
const input = require("fs").readFileSync(0).toString().trim().split(" ").map(Number);

const [N, M] = input;

const visited = new Array(N);
const output = [];
let result = "";

const DFS = (index, count) => {
  if (count === M) {
    result += `${output.join(" ")}\n`;
    return;
  }

  for (let i = index; i < N; i++) {
    if (visited[i] === true) continue;

    visited[i] = true;
    output.push(i + 1);
    DFS(i, count + 1);
    output.pop();
    visited[i] = false;
  }
};

DFS(0, 0);
console.log(result.trim());

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.py
def dfs(start):
    if len(stack) == M:
        print(' '.join(map(str, stack)))
        return

    for i in range(start, N + 1):
        if visited[i]:
            continue

        visited[i] = 1
        stack.append(i)
        dfs(i + 1)
        stack.pop()
        visited[i] = 0

N, M = map(int, input().split())
stack = []
visited = [0] * (N + 1)

dfs(1)